### PR TITLE
Issue #2576: Fix failing admin bar test.

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/2576

### CODE REVIEW

File: core/modules/user/user.module [hunk 1]
```diff
@@ -497,8 +497,9 @@ function user_permission() {
       'title' => t('Administer users'),
       'restrict access' => TRUE,
     ),
+    // @todo change this to assign roles
```
Comment (quicksketch) [inline core/modules/user/user.module]: Remove this `@todo`

File: core/modules/user/user.module [hunk 2]
```diff
@@ -2263,21 +2253,30 @@ function user_roles($members_only = FALSE, $permission = NULL, $full_objects = F
 }
 
 /**
- * Helper for user_roles().
+ * Checks if access should be restricted.
```
Comment (quicksketch) [inline core/modules/user/user.module]: A more verbose description should be used here. What makes a role "restricted"? The docblock should describe it without requiring the developer to read the source code.

File: core/modules/user/user.module [hunk 3]
```diff
@@ -497,8 +497,9 @@ function user_permission() {
       'title' => t('Administer users'),
       'restrict access' => TRUE,
     ),
+    // @todo change this to assign roles
     'administer user roles' => array(
```
Comment (quicksketch) [inline core/modules/user/user.module]: Hm, perhaps the machine name for this permission to match the description?

File: core/modules/user/user.module [hunk 4]
```diff
@@ -2773,7 +2772,7 @@ function user_action_info() {
       continue;
     }
     if (!user_access('administer user roles') && !user_access('administer permissions')) {
```
Comment (quicksketch) [inline core/modules/user/user.module]: Adding a comment here would also be a good thing.

**Review by quicksketch — CHANGES_REQUESTED** [review]

File: core/modules/user/user.module [hunk 5]
```diff
@@ -2261,21 +2252,30 @@ function user_roles($members_only = FALSE, $permission = NULL, $full_objects = F
 }
 
 /**
- * Helper for user_roles().
+ * Checks if access should be restricted.
+ *
+ * @param string $role
+ *   User Role.
```
Comment (quicksketch) [inline core/modules/user/user.module]: :face_with_head_bandage:

**Comment by quicksketch** [comment]
We could also use an update hook that adds the `assign roles` permission to all roles that have the `administer permission` permission, otherwise updating sites are going to have any roles that allow assigning permissions (even the so-called "admin role").

File: core/modules/user/user.module (hunk 2 above)
Comment (jenlampton) [reply core/modules/user/user.module]: How about "Checks a role to see if any of its permissions have the 'restrict access' flag set."

File: core/modules/user/user.module (hunk 5 above)
Comment (jenlampton) [reply core/modules/user/user.module]: I'm not sure `A user role object.` is any better.

**Comment by jenlampton** [comment]
I added an update hook and betterified some of the docs.

File: core/modules/user/user.install [hunk 6]
```diff
@@ -1138,5 +1137,20 @@ function user_update_1012() {
...
+    $names = config_get_names_with_prefix('user.role.');
+    foreach ($names as $config_name) {
+      $config = config($config_name);
+      $permissions = $config->get('permissions');
+      if (in_array('administer permissions', $permissions)) {
+        $permissions[] = 'assign roles';
+        $config->save();
+      }
+    }
+}
```
Comment (quicksketch) [inline core/modules/user/user.install]: There seems to be an extra space of indentation in this code block.

Comment (jenlampton) [reply core/modules/user/user.install]: can fix!

File: core/modules/user/user.install [hunk 7]
```diff
@@ -1138,5 +1137,20 @@ function user_update_1012() {
...
+ * Grant accounts with 'administer permissions' the new 'assign roles' permission.
+ */
+function user_update_1013() {
+  $names = config_get_names_with_prefix('user.role.');
+  foreach ($names as $config_name) {
+    $config = config($config_name);
+    $permissions = $config->get('permissions');
+    if (in_array('administer permissions', $permissions)) {
+      $permissions[] = 'assign roles';
+      $config->save();
```
Comment (quicksketch) [inline core/modules/user/user.install]: This Update hook doesn't seem to work. It's missing the assignment of $permissions back into $config->set('permissions', $permissions).

Comment (jenlampton) [reply core/modules/user/user.install]: Hah!

**Pull request merged**